src/common/users.ts

```diff
@@ -113,35 +113,44 @@ export interface ReadingDaysArgs {
   dateRange: DateRange;
 }
 
-export const getUserReadingDays = (
+export const getUserReadingTags = (
   con: Connection,
   { userId, dateRange: { start, end }, limit = 8 }: ReadingDaysArgs,
 ): Promise<TagsReadingStatus[]> => {
+  const timezone = `COALESCE(u.timezone, 'utc')`;
   return con.query(
     `
-    with filtered_view as (
-      select  *, CAST(v."timestamp" at time zone COALESCE(u.timezone, 'utc') AS DATE) as day
-      from    "view" v
-      inner   join "user" u
-      on      u."id" = v."userId"
-
-      where   u."id" = $1
-      and     "timestamp" >= $2
-      and     "timestamp" < $3
-    )
-    select  *,
-            (select count(DISTINCT day) from filtered_view) as total,
-            tags."readingDays" * 1.0 / (select count(DISTINCT day) from filtered_view) as percentage
-    from (
-      select pk.keyword as tag, count(DISTINCT day) as "readingDays"
-      from filtered_view v
-      inner join post_keyword pk on v."postId" = pk."postId" and pk.status = 'allow'
-      where pk.keyword != 'general-programming'
-      group by pk.keyword
-    ) as tags
-    order by tags."readingDays" desc
-    limit $4;
-  `,
+      with filtered_view as (
+        select *,
+               CAST(v."timestamp" at time zone ${timezone} AS DATE) as day
+      from "view" v
+        inner join "user" u
+      on u."id" = v."userId"
+
+      where u."id" = $1
+        and "timestamp" >= date_trunc('week'
+          , timezone(${timezone}
+          , $2))
+        and "timestamp"
+          < date_trunc('week'
+          , timezone(${timezone}
+          , $3)) + interval '7 days'
```

Comment:
Does this not break the year function on the profile page?

Reply:
When I did the manual QA I mentioned that it's already broken because it doesn't take into account the full week. So I thought I fixed that now but I was confused with the reading rank. It's not relevant for the tags. I'll have to look for another solution.
What bothers me actually the most is that we have several ways in the same function to calculate the date range. I'm not a fan and we might want to address it in the near future as it can cause misalignments between the different queries

Reply:
Good point we should actually combine these date based queries into one generic function that takes the timezone into account.